### PR TITLE
Track View: Major fix for GHI# 11500

### PR DESCRIPTION
**Pull request by Aleks-Starykh-TR**
## What does this PR do?
Closes https://github.com/o3de/o3de/issues/11500
Closes https://github.com/o3de/o3de/issues/17372

* Fixes the ambiguous warning `"Can't get component (type: EditorSequenceComponent, addr: ?) entity ID as it is not attached to an entity yet!"` in the `Component::GetEntityId()` generated in the `EditorSequenceComponent::RemoveEntityToAnimate(()` while destroying an instance when sanitizing a prefab DOM;
* * No side-effects.

--
* Fixes the ambiguous warning `"The dirty entity has no owning instance."` in the `PrefabPublicHandler::GenerateUndoNodesForEntityChangeAndUpdateCache()`  generated while destroying an instance when sanitizing a prefab DOM; 
* * This also meant that undo state was broken, which could spoil prefab template operations;
* * This occurred only when destructor for `EditorSequenceComponent`: was called before the destructor of the linked `EditorSequenceAgentComponent`; note that order of entities / components destruction is randomized, - I believe this depends on randomly generated Entity Ids and Component Ids and hence their positions in hash-tables, - for example I have two seemingly same level prefabs, differing only in entity aliases and components, which are randomly generated, and one of these was misbehaving;
* * So, in randomized destruction order this also broke the prefab loading / saving pipeline due to the `EditorSequenceComponent::RemoveEntityToAnimate()` was called second time from undo chain and thus was calling the `EditorSequenceAgentComponent::DisconnectSequence()` second time . 
* * The first variant of a fix was removing  the `CAnimAzEntityNode` containing the `removedEntityId` after calling the `RemoveEntityToAnimate()` from the EditorSequenceComponent destructor, and thus disconnecting this sequence component with that `removedEntityId` entity and destroying the `EditorSequenceAgentComponent` , yet this could introduce misbehaviour in the governing code which neglects calling  AddEntityToAnimate() / RemoveEntityToAnimate()
* * The final fix is in temporary suppressing undo / redo during the `EditorSequenceAgentComponent` destruction in the `EditorSequenceAgentComponent::DisconnectSequence()`, as it this component is constructed not through the normal pipeline (designed for direct user actions evaluation). In order to temporary suppress undo / redo for this specific operation the commit https://github.com/o3de/o3de/pull/18538/commits/191e5cff77d2c958ca09414ee2f04bc829072d66 :
* * * adds the new API to maintain "ignored entities list",
* * * adds checking this list when attempting to add dirty entities,
* * * and then uses this new API bracketing the the 'normal' codepath for the` EditorSequenceAgentComponent` destruction in the `EditorSequenceAgentComponent::DisconnectSequence()`.

--
* Also fixes the crash on invalid pointer in `CAnimSequence::RemoveNode()` moving node destruction down below the `node` pointer  usage.
* * Prior to this fix, calling the `CAnimSequence::RemoveNode(IAnimNode* node)` code first destructed the object erasing a `vector` element with the `intrusive_ptr<IAnimNode>`, and then attempted to use the now invalid `node` pointer;
* * The commit https://github.com/o3de/o3de/pull/18538/commits/a991923856cf7fb7443c56b1e57ff0e61ee31b82 also adds a "TODO" comment  to the identical code implementation of the `CUiAnimSequence::RemoveNode()`, changing and testing of which is out of scope for the current WIP.
* * This method was used in the previous version of the fix described above now used in a fix described below;
--

## How was this PR tested?
Repro steps of the issue https://github.com/o3de/o3de/issues/11500 with the `Camera2` and the `Seq` entities repeated (numerously) under Windows:
a `TestLevel` with a `Camera2` and  a `Seq` entities containing (correctly linked `EditorSequenceAgentComponent` and `EditorSequenceComponent` respectively),  - are now correctly saved an loaded.
Detached sequence entities were never noted in this scenario.

### What is not done yet?
* When instantiating a prefab which was previously (with earlier code versions) saved with a spoiled (empty) "Parent Entity" sentence in the `TransformComponent` of a sequence entity, such prefab is still loaded wrongly;
I believe fixing this would be helpful, as in most practical cases the AssetBuilder will rebuild a buggy `prefab` into a correct `spawnable` - see PR #18548.
* The problem also noted with incorrect structure of an in-memory template for a prefab recently created from a pair of `Camera2` and  a `Seq` entities in a level:
* * If the newly created prefab is deleted from the level, and then re-instantiated, it is instantiated with the "torn-off" `Seq` entity, because the cached in-memory DOM template is used, and this one is somehow asynchronously spoiled (the `Seq` entity forgets the correct container entity alias in the "Parent Entity" statement).
* * If a level is saved / reloaded or another level is loaded (thus releasing cached  in-memory templates), the newly created prefab is instantiated correctly.

### CODE REVIEW

File: Gems/Maestro/Code/Source/Cinematics/AnimSequence.cpp [hunk 1]
```diff
@@ -387,13 +387,14 @@ void CAnimSequence::RemoveNode(IAnimNode* node, bool removeChildRelationships)
...
         if (node == m_nodes[i].get())
         {
-            m_nodes.erase(m_nodes.begin() + i);
-
             if (node->NeedToRender())
             {
                 RemoveNodeNeedToRender(node);
             }
 
+            // TODO : move destruction after usage in CUiAnimSequence::RemoveNode() too ?
```
Comment (guillaume-haerinck) [inline Gems/Maestro/Code/Source/Cinematics/AnimSequence.cpp:395]: Put the comment rather in CUiAnimSequence::RemoveNode() and use the intial issue such as 
// TODO #11500 Moving node destruction after usage below might help to fix sequence entity spawning, but needs more investigation

File: Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp [hunk 2]
```diff
@@ -234,10 +234,35 @@ namespace Maestro
...
         // Notify the SequenceAgentComponent that we're disconnecting from it
         Maestro::SequenceAgentComponentRequestBus::Event(ebusId, &Maestro::SequenceAgentComponentRequestBus::Events::DisconnectSequence);
+
+        // A linked animation agent in a referenced entity disconnects with the above code.
+        // Removing the node disables next calls to EditorSequenceAgentComponent::DisconnectSequence()
+        // and thus fixes ambiguous warnings on missing owner instance in ToolsApplication::CreateUndosForDirtyEntities()
+        // while destroying a prefab instance
+        if (m_sequence.get())
+        {
+            for (int i = m_sequence->GetNodeCount(); --i >= 0;)
```
Comment (guillaume-haerinck) [inline Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp:257]: why the decrement in the condition ? Its not working with `i >=0; --i)' ?

File: Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp [hunk 3]
```diff
@@ -234,10 +234,35 @@ namespace Maestro
...
+            {
+                IAnimNode* animNode = m_sequence->GetNode(i);
+                if (animNode->GetType() == AnimNodeType::AzEntity && animNode->GetAzEntityId() == removedEntityId)
+                {
+                    m_sequence->RemoveNode(animNode, true);
```
Comment (guillaume-haerinck) [inline Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp:262]: maybe put the 'true' in a constexpr bool descriptiveName = true; or use /*varName*/ in the ()

**Review by guillaume-haerinck — COMMENTED** [review]
Good job, its crazy to think that this got broken for so long. When I did fixes in the ScriptCanvas logger and wasn't able to figure out the code, using breakpoints in Lumberyard helped a ton to understand what broke it/what would be the proper fix, might be worth doing that here as well

File: Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp (hunk 2 above)
Comment (Aleks-Starykh-TR) [reply Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp:257]: Yes, it would work as You propose, yet this would be
`for (int i = m_sequence->GetNodeCount() - 1; i >= 0; --i)`
implying one extra CPU substaraction op  "`- 1`" (in general neglecting probable compiler optimizer changes).
Truthly, this line was copy / pasted from `EditorSequenceComponent::~EditorSequenceComponent() Line:54`, so I regarded this style (a sort of an old school C-style source code optimization) as recognized and approved by Community, and no opposing statements found in "O3DE C++ Coding Standards".

File: Gems/Maestro/Code/Source/Cinematics/AnimSequence.cpp (hunk 1 above)
Comment (Aleks-Starykh-TR) [reply Gems/Maestro/Code/Source/Cinematics/AnimSequence.cpp]: Thanks, this surely makes great sense.
Moved the TODO comment to the CUiAnimSequence::RemoveNode() in commit  https://github.com/o3de/o3de/pull/18538/commits/c332944c92542710caac1015263422121fe5dc0e.

File: Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp (hunk 2 above)
Comment (Aleks-Starykh-TR) [reply Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp:285]: Anyway, with commit https://github.com/o3de/o3de/pull/18538/commits/c5c29f85d1c3bab95a1e39bc0b445fad5520f3c2 I've commented out this loop, reverting to safer (minimal) fix for the bug generating ambiguous warning and disrupting a (level) prefab loading.
The problem is that the most sensefull from the architectural point of view change may have side-effects with the current code implementation (when governing code removing an animated sequnce link forgets to re-add this later before using this animation). 
Until a final GHI resolution I am keeping the proposed code, commenting it out with (with `#if 0` ... `#endif // 0` brackets.
The relevant code snippets tested by me are thoroughly commented with "TODO (AnimSequence linking)" tags.

File: Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp (hunk 3 above)
Comment (Aleks-Starykh-TR) [reply Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp]: Thanks, makes sence for readability.
In commit https://github.com/o3de/o3de/pull/18538/commits/c5c29f85d1c3bab95a1e39bc0b445fad5520f3c2 I am adding the named bool constant as the 2nd argument both in this commented out code snippet and in more safe minimal fix code.

**Comment by guillaume-haerinck** [comment]
A bit too much TODO code for my taste but given that without that the trackview is almost completely broken it make sense to bring the feature back, thanks for looking at it !

**Review by guillaume-haerinck — APPROVED** [review]

**Review by lsemp3d — APPROVED** [review]

File: Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp [hunk 4]
```diff
@@ -234,10 +249,50 @@ namespace Maestro
...
+        // in the component dtor above, where this method is called directly.
+        // Removing the node disables any following ambiguous bus calls to the
+        // EditorSequenceAgentComponent::DisconnectSequence() in this component,
+        // as the corresponding agent component is now detached and "dead"
+        // (please see comments in the EditorSequenceAgentComponent::DisconnectSequence()).
+        // Although it seems that destroying the now ambiguous node here would be architecturally right,
+        // this may interfere with the current implementation of the governing code.
+        // In ideal world the governing code calling RemoveEntityToAnimate() is then to call the AddEntityToAnimate()
+        // in case the Sequence <-> SequenceAgent link is further used.
+        #if 0
```
Comment (lsemp3d) [inline Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp:278]: Will this be changed in a different PR? any reason to keep it disabled?

Comment (Aleks-Starykh-TR) [reply Gems/Maestro/Code/Source/Components/EditorSequenceComponent.cpp]: @lsemp3d @guillaume-haerinck 
The final fix for the issue does not need removing the CAnimAzEntityNode, so I deleted most of the above comments, leaving a TODO in a single place - because I believe this architectural inconsistency deserves mentioning.

The PR description above updated.
Code is ready for merging.
Thanks for your patience.

File: Gems/Maestro/Code/Source/Components/EditorSequenceAgentComponent.cpp [hunk 5]
```diff
@@ -165,8 +173,83 @@ namespace Maestro
         AZ::EntityId curEntityId = GetEntityId();
 
         // remove this SequenceAgent from this entity if no sequenceComponents are connected to it
-        AzToolsFramework::EntityCompositionRequestBus::Broadcast(&AzToolsFramework::EntityCompositionRequests::RemoveComponents, AZ::Entity::ComponentArrayType{this});        
+        {
+            AZ_Trace("EditorSequenceAgentComponent", "DisconnectSequence(): removing self from entity %s, %s.",
+                curEntityId.ToString().c_str(), entity->GetName().c_str());
+            // The normal call for a component would be:
+            // AzToolsFramework::EntityCompositionRequestBus::Broadcast(&AzToolsFramework::EntityCompositionRequests::RemoveComponents, AZ::Entity::ComponentArrayType{this});
+            // but this one is not used, as this component is created abnormally via user actions in EditorSequenceComponent.
```
Comment (guillaume-haerinck) [inline Gems/Maestro/Code/Source/Components/EditorSequenceAgentComponent.cpp:181]: Do you know the reason behind the fact this components are created abnormaly ? Because it sounds like that it shouldn't be if we have to 'hack' the destruction if the construction was also 'hacked'. Again I'm not too familiar with the trackview architecture in O3DE (is there some reviewers who 'owns' trackview ?)

Comment (Aleks-Starykh-TR) [reply Gems/Maestro/Code/Source/Components/EditorSequenceAgentComponent.cpp:181]: New Entity Ids, new Entity Aliases, new Component Ids and new Instance Ids are generated randomly, and then used as keys into unordered maps.
As an `EditorSequenceAgentComponent` is not created directly by user action, but as a result of a `EditorSequenceComponent` manipulation, - it does not need a local undo stack for its construction / destruction,  - and this local undo stack became broken when `EditorSequenceComponent` `dtor` was call before the `EditorSequenceAgentComponent` `dtor`.
One of the reasons I failed to catch this during days-and-days-and-days debugging line-after-line, is that the `EditorSequenceAgentComponent` misses explicit `ctor` and `dtor` I could put breakpoints on. Next time I'll add all missing methods and put logs into these, as I've done here (but then cleaned-up debugging code).

Does this answer you question?

Comment (guillaume-haerinck) [reply Gems/Maestro/Code/Source/Components/EditorSequenceAgentComponent.cpp:181]: Yeah totally thanks for the explanation 👍 Twice in my career I've seen people working fulltime and struggling on a trackview feature, it does sounds like a very complex area to debug

Comment (guillaume-haerinck) [reply Gems/Maestro/Code/Source/Components/EditorSequenceAgentComponent.cpp:181]: I'm a tad worried that this custom spawn and destroy logic will become a pain to maintain, as you'll want to have any modification done in the 'normal' logic. If the issue comes from the undo, isn't it possible to use the 'normal' codepath but asking for undo to be disabled during spawn/unspawn ?

Comment (Aleks-Starykh-TR) [reply Gems/Maestro/Code/Source/Components/EditorSequenceAgentComponent.cpp]: > isn't it possible to use the 'normal' codepath but asking for undo to be disabled during spawn/unspawn ?

Managing undo / redo state is (rightly) encapsulated, and I won't dare to mess with the inner undo / redo logic, 
I also believe that undo / redo may work during spawn/despawn when evaluating patches.

So to temporary suppress undo / redo for specific operations I've
added the new API to maintain "ignored entities list",
added checking this list when attempting to add dirty entities,
and then used this new API bracketing the the 'normal' codepath for the `EditorSequenceAgentComponent` destruction in the `EditorSequenceAgentComponent::DisconnectSequence()` (thus reverting to using the original 'normal' codepath).

Please see commit https://github.com/o3de/o3de/pull/18538/commits/191e5cff77d2c958ca09414ee2f04bc829072d66.

**Review by guillaume-haerinck — APPROVED** [review]

**Comment by Aleks-Starykh-TR** [comment]
@lsemp3d : You have previously approved it, but I've re-requested your review after code changes.

**Review by nick-l-o3de — APPROVED** [review]
I guess the alternative would have been to call RemoveDirtyEntity but not having it enter the undo stack and be captured in the first place really is preferable.  The same API might be useful later for creating temporary visualization entities that must not be saved or persisted.

**Comment by Aleks-Starykh-TR** [comment]
@nick-l-o3de 
> I guess the alternative would have been to call RemoveDirtyEntity but not having it enter the undo stack and be captured in the first place really is preferable. The same API might be useful later for creating temporary visualization entities that must not be saved or persisted.

The problem is that calling `RemoveDirtyEntity` before the standard code-flow request to the  `AzToolsFramework::EntityCompositionRequests::RemoveComponents` does not help, as the Entity is added to a dirty list deep inside this standard code-flow.

**Comment by nick-l-o3de** [comment]
Excellent work on this PR

**Review by lsemp3d — APPROVED** [review]
